client/engine/chainservice/eth_chainservice.go

```diff
@@ -89,16 +91,23 @@ func newEthChainService(chain ethChain, na *NitroAdjudicator.NitroAdjudicator,
 	ctx, cancelCtx := context.WithCancel(context.Background())
 
 	// Use a buffered channel so we don't have to worry about blocking on writing to the channel.
-	ecs := EthChainService{chain, na, naAddress, caAddress, vpaAddress, txSigner, make(chan Event, 10), logger, ctx, cancelCtx}
+	ecs := EthChainService{chain, na, naAddress, caAddress, vpaAddress, txSigner, make(chan Event, 10), logger, ctx, cancelCtx, sync.WaitGroup{}}
 	errChan, err := ecs.subscribeForLogs()
 	// TODO: Return error from chain service instead of panicking
+	ecs.wg.Add(1)
 	go func() {
-		for err := range errChan {
-			// Print to STDOUT in case we're using a noop logger
-			fmt.Println(err)
-			ecs.logger.Fatal().Err(err)
-			// Manually panic in case we're using a logger that doesn't call exit(1)
-			panic(err)
+		for {
+			select {
+			case <-ecs.ctx.Done():
+				ecs.wg.Done()
+				return
+			case err := <-errChan:
+				// Print to STDOUT in case we're using a noop logger
+				fmt.Println(err)
+				ecs.logger.Fatal().Err(err)
+				// Manually panic in case we're using a logger that doesn't call exit(1)
+				panic(err)
+			}
```

Comment:
Like encapsulate 
```
// Print to STDOUT in case we're using a noop logger
				fmt.Println(err)
				ecs.logger.Fatal().Err(err)
				// Manually panic in case we're using a logger that doesn't call exit(1)
				panic(err)
```

into a `handleError` function?

Reply:
Yeah but I mean the whole for/ select loop

Reply:
https://github.com/statechannels/go-nitro/pull/1343/commits/e5622ef27f67c621518620c63a37b1427a677094